src/circuit/gadgets/constraints.rs

```diff
@@ -698,6 +799,172 @@ mod tests {
 
     proptest! {
 
+        #[test]
+        fn test_enforce_equal((a, b) in any::<(FWrap<Fr>, FWrap<Fr>)>()) {
+            prop_assume!(a != b);
+
+            let test_a_b = |a, b| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(a)).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(b)).unwrap();
+                let _ = enforce_equal(&mut cs, || "enforce equal", &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), a==b);
+            };
+
+            // positive
+            test_a_b(a.0, a.0);
+
+            // negative
+            test_a_b(a.0, b.0);
+        }
+
+        #[test]
+        fn test_enforce_equal_zero(a in (1u64..42)) {
```

Comment:
Maybe we want a 'non-zero field element' strategy. And a 'non-one field element'. Just thinking out loud about how we can evolve the approach.

Reply:
@huitseeker Can we have such strategies?

Reply:
I think I remember you can filter.